src/utils/masking.py

```diff
@@ -0,0 +1,115 @@
+import random
+from typing import List, Optional, Union
+
+import numpy as np
+
+
+class SpanMaskingGenerator:
+    """
+    Generator class that yields span masks
+
+    Args:
+        num_patches (`int`):
+            The total number of images patches
+        num_masking_patches (`int`, defaults to 1):
+            The number of patches to be masked out. Typically determined by the masking ratio
+        max_span_length (`int`, defaults to 6):
+            The maximum number of consecutive masked patches
+        spacing (`Union[int, str]`, default to 0):
+            The number of non-masked patches in between consecutive masked spans. Can either be an integer value,
+            in which case the spacing is fixed, or can be set to "span" in which case the spacing is dynamic such
+            that on both sides of a masked span of length N patches, there will be N non-masked patches. Note that
+            larger spacing makes it harder to greedily sample masks satisfying these constraints which can slow down
+            masking and also cause the algorithm to terminate with a smaller mask than specified. In case of the
+            latter, PIXEL randomly masks additional patches until the specified masking ratio is reached.
+
+            These are the recommended settings:
+             - For masking ratio <= 0.4 use "span" mode.
+             - For ratios between 0.4 and 0.7 set spacing to 1.
+             - For higher, set spacing to 0
+    """
+
+    def __init__(
+        self,
+        num_patches: int,
+        num_masking_patches: int = 1,
+        max_span_length: int = 6,
+        spacing: Union[int, str] = 0,
+        cumulative_span_weights: Optional[List[float]] = None,
+    ):
+
+        self.num_patches = num_patches
+        self.num_masking_patches = num_masking_patches
+
+        self.max_span_length = max_span_length
+        self.spacing = spacing
+        assert spacing == "span" or isinstance(spacing, int)
+
+        self.span_range = range(1, max_span_length + 1)
+        self.cumulative_span_weights = cumulative_span_weights
+
+    def _mask(self, mask, max_mask_patches):
+        delta = 0
+        # Lower number of attempts will speed up mask generation but might cause a lot fewer patches to be masked
+        # than desired, particularly for high masking ratios
+        for attempt in range(100):
+            # Randomly select span length within specified range
+            span = random.choices(self.span_range, cum_weights=self.cumulative_span_weights, k=1)[0]
+            if span < self.num_patches:
```

Comment:
Am I right that if we chose the span size from the range(1, max_mask_patches + 1) one line above, than we don't need the check here?